File: RarcFilesystem.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;

namespace Whitehole
{
    public class RarcFilesystem : FilesystemBase
    {
        public RarcFilesystem(FileBase file)
        {
            m_File = file;
            m_File.Stream = new Yaz0Stream(m_File.Stream);
            m_File.BigEndian = true;

            m_File.Stream.Position = 0;
            uint tag = m_File.Reader.ReadUInt32();
            if (tag != 0x52415243) throw new Exception("File isn't a RARC (tag 0x" + tag.ToString("X8") + ", expected 0x52415243)");

            m_File.Stream.Position = 0xC;
            m_FileDataOffset = m_File.Reader.ReadUInt32() + 0x20;
            m_File.Stream.Position = 0x20;
            m_NumDirNodes = m_File.Reader.ReadUInt32();
            m_DirNodesOffset = m_File.Reader.ReadUInt32() + 0x20;
            m_File.Stream.Position += 0x4;
            m_FileEntriesOffset = m_File.Reader.ReadUInt32() + 0x20;
            m_File.Stream.Position += 0x4;
            m_StringTableOffset = m_File.Reader.ReadUInt32() + 0x20;

            m_DirEntries = new Dictionary<uint, DirEntry>();
            m_FileEntries = new Dictionary<uint, FileEntry>();

            DirEntry root = new DirEntry();
            root.ID = 0;
            root.ParentDir = 0xFFFFFFFF;

            m_File.Stream.Position = m_DirNodesOffset + 0x6;
            uint rnoffset = m_File.Reader.ReadUInt16();
            m_File.Stream.Position = m_StringTableOffset + rnoffset;
            root.Name = m_File.ReadString();
            root.FullName = "/" + root.Name;

            m_DirEntries.Add(0, root);

            for (uint i = 0; i < m_NumDirNodes; i++)
            {
                DirEntry parentdir = m_DirEntries[i];

                m_File.Stream.Position = m_DirNodesOffset + (i * 0x10) + 10;

                ushort numentries = m_File.Reader.ReadUInt16();
                uint firstentry = m_File.Reader.ReadUInt32();

                for (uint j = 0; j < numentries; j++)
                {
                    uint entryoffset = m_FileEntriesOffset + ((j + firstentry) * 0x14);
                    m_File.Stream.Position = entryoffset;

                    uint fileid = m_File.Reader.ReadUInt16();
                    m_File.Stream.Position += 4;
                    uint nameoffset = m_File.Reader.ReadUInt16();
                    uint dataoffset = m_File.Reader.ReadUInt32();
                    uint datasize = m_File.Reader.ReadUInt32();

                    m_File.Stream.Position = m_StringTableOffset + nameoffset;
                    string name = m_File.ReadString();
                    if (name == "." || name == "..") continue;

                    string fullname = parentdir.FullName + "/" + name;

                    if (fileid == 0xFFFF)
                    {
                        DirEntry d = new DirEntry();
                        d.EntryOffset = entryoffset;
                        d.ID = dataoffset;
                        d.ParentDir = i;
                        d.NameOffset = nameoffset;
                        d.Name = name;
                        d.FullName = fullname;

                        m_DirEntries.Add(dataoffset, d);
                    }
                    else
                    {
                        FileEntry f = new FileEntry();
                        f.EntryOffset = entryoffset;
                        f.ID = fileid;
                        f.ParentDir = i;
                        f.NameOffset = nameoffset;
                        f.DataOffset = dataoffset;
                        f.DataSize = datasize;
                        f.Name = name;
                        f.FullName = fullname;

                        m_FileEntries.Add(fileid, f);
                    }
                }
            }
        }

        public override void Close()
        {
            m_File.Close();
        }


        public override bool DirectoryExists(string directory)
        {
            DirEntry dir = m_DirEntries.Values.FirstOrDefault(de => de.FullName.ToLower() == directory.ToLower());
            return dir != null;
        }

        public override string[] GetDirectories(string directory)
        {
            DirEntry dir = m_DirEntries.Values.FirstOrDefault(de => de.FullName.ToLower() == directory.ToLower());
            if (dir == null) return null;
            IEnumerable<DirEntry> subdirs = m_DirEntries.Values.Where(de => de.ParentDir == dir.ID);

            List<string> ret = new List<string>();
            foreach (DirEntry de in subdirs) ret.Add(de.Name);
            return ret.ToArray();
        }


        public override bool FileExists(string filename)
        {
            FileEntry file = m_FileEntries.Values.FirstOrDefault(fe => fe.FullName.ToLower() == filename.ToLower());
            return file != null;
        }

        public override string[] GetFiles(string directory)
        {
            DirEntry dir = m_DirEntries.Values.FirstOrDefault(de => de.FullName.ToLower() == directory.ToLower());
            if (dir == null) return null;
            IEnumerable<FileEntry> files = m_FileEntries.Values.Where(fe => fe.ParentDir == dir.ID);

            List<string> ret = new List<string>();
            foreach (FileEntry fe in files) ret.Add(fe.Name);
            return ret.ToArray();
        }

        public override FileBase OpenFile(string filename)
        {
            FileEntry file = m_FileEntries.Values.FirstOrDefault(fe => fe.FullName.ToLower() == filename.ToLower());
            if (file == null) return null;

            return new RarcFile(this, file.ID);
        }


        // support functions for RarcFile
        public byte[] GetFileContents(RarcFile file)
        {
            FileEntry fe = m_FileEntries[file.ID];

            m_File.Stream.Position = m_FileDataOffset + fe.DataOffset;
            return m_File.Reader.ReadBytes((int)fe.DataSize);
        }

        public void ReinsertFile(RarcFile file)
        {
            FileEntry fe = m_FileEntries[file.ID];

            uint fileoffset = m_FileDataOffset + fe.DataOffset;
            int oldlength = (int)fe.DataSize;
            int newlength = (int)file.Stream.Length;
            int delta = newlength - oldlength;

            if (newlength != oldlength)
            {
                m_File.Stream.Position = fileoffset + oldlength;
                byte[] tomove = m_File.Reader.ReadBytes((int)(m_File.Stream.Length - m_File.Stream.Position));

                m_File.Stream.Position = fileoffset + newlength;
                m_File.Stream.SetLength(m_File.Stream.Length + delta);
                m_File.Writer.Write(tomove);

                fe.DataSize = (uint)newlength;
                m_File.Stream.Position = fe.EntryOffset + 0xC;
                m_File.Writer.Write(fe.DataSize);

                foreach (FileEntry tofix in m_FileEntries.Values)
                {
                    if (tofix.ID == fe.ID) continue;
                    if (tofix.DataOffset < (fe.DataOffset + oldlength)) continue;

                    tofix.DataOffset = (uint)(tofix.DataOffset + delta);
                    m_File.Stream.Position = tofix.EntryOffset + 0x8;
                    m_File.Writer.Write(tofix.DataOffset);
                }
            }

            m_File.Stream.Position = fileoffset;
            file.Stream.Position = 0;
            byte[] data = file.Reader.ReadBytes(newlength);
            m_File.Writer.Write(data);

            m_File.Flush();
        }


        private class FileEntry
        {
            public uint EntryOffset;

            public uint ID;
            public uint NameOffset;
            public uint DataOffset;
            public uint DataSize;

            public uint ParentDir;

            public string Name;
            public string FullName;
        }

        private class DirEntry
        {
            public uint EntryOffset;

            public uint ID;
            public uint NameOffset;

            public uint ParentDir;

            public string Name;
            public string FullName;
        }


        private FileBase m_File;

        private uint m_FileDataOffset;
        private uint m_NumDirNodes;
        private uint m_DirNodesOffset;
        private uint m_FileEntriesOffset;
        private uint m_StringTableOffset;

        private Dictionary<uint, FileEntry> m_FileEntries;
        private Dictionary<uint, DirEntry> m_DirEntries;
    }


    public class RarcFile : FileBase
    {
        public RarcFile(RarcFilesystem fs, uint id)
        {
            m_FS = fs;
            m_ID = id;

            byte[] buffer = m_FS.GetFileContents(this);
            Stream = new MemoryStream(buffer.Length);
            Writer.Write(buffer);
        }

        public override void Flush()
        {
            Stream.Flush();
            m_FS.ReinsertFile(this);
        }


        private RarcFilesystem m_FS;
        private uint m_ID;

        public uint ID { get { return m_ID; } }
    }
}

File: src/whitehole/fileio/RarcFilesystem.java
/*
    Copyright 2012 The Whitehole team

    This file is part of Whitehole.

    Whitehole is free software: you can redistribute it and/or modify it under
    the terms of the GNU General Public License as published by the Free
    Software Foundation, either version 3 of the License, or (at your option)
    any later version.

    Whitehole is distributed in the hope that it will be useful, but WITHOUT ANY 
    WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS 
    FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along 
    with Whitehole. If not, see http://www.gnu.org/licenses/.
*/

package whitehole.fileio;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.*;

public class RarcFilesystem implements FilesystemBase
{
    public RarcFilesystem(FileBase _file) throws IOException
    {
        file = new Yaz0File(_file);
        file.setBigEndian(true);

        file.position(0);
        int tag = file.readInt();
        if (tag != 0x52415243) 
            throw new IOException(String.format("File isn't a RARC (tag 0x%1$08X, expected 0x52415243)", tag));

        file.position(0xC);
        int fileDataOffset = file.readInt() + 0x20;
        file.position(0x20);
        int numDirNodes = file.readInt();
        int dirNodesOffset = file.readInt() + 0x20;
        int numFileEntries = file.readInt();
        int fileEntriesOffset = file.readInt() + 0x20;
        file.skip(0x4);
        int stringTableOffset = file.readInt() + 0x20;
        unk38 = file.readInt();

        dirEntries = new LinkedHashMap<>(numDirNodes);
        fileEntries = new LinkedHashMap<>(numFileEntries);

        DirEntry root = new DirEntry();
        root.parentDir = null;

        file.position(dirNodesOffset + 0x6);
        int rnoffset = file.readShort();
        file.position(stringTableOffset + rnoffset);
        root.name = file.readString("ASCII", 0);
        root.fullName = "/" + root.name;
        root.tempID = 0;

        dirEntries.put("/", root);
        for (int i = 0; i < numDirNodes; i++)
        {
            DirEntry parentdir = null;
            for (DirEntry de : dirEntries.values())
            {
                if (de.tempID == i)
                {
                    parentdir = de;
                    break;
                }
            }

            file.position(dirNodesOffset + (i * 0x10) + 10);

            short numentries = file.readShort();
            int firstentry = file.readInt();
            for (int j = 0; j < numentries; j++)
            {
                int entryoffset = fileEntriesOffset + ((j + firstentry) * 0x14);
                file.position(entryoffset);

                file.skip(4);
                int entrytype = file.readShort() & 0xFFFF;
                int nameoffset = file.readShort() & 0xFFFF;
                int dataoffset = file.readInt();
                int datasize = file.readInt();

                file.position(stringTableOffset + nameoffset);
                String name = file.readString("ASCII", 0);
                if (name.equals(".") || name.equals("..")) continue;
                
                String fullname = parentdir.fullName + "/" + name;

                if (entrytype == 0x0200)
                {
                    DirEntry d = new DirEntry();
                    d.parentDir = parentdir;
                    d.name = name;
                    d.fullName = fullname;
                    d.tempID = dataoffset;

                    dirEntries.put(pathToKey(fullname), d);
                    parentdir.childrenDirs.add(d);
                }
                else
                {
                    FileEntry f = new FileEntry();
                    f.parentDir = parentdir;
                    f.dataOffset = fileDataOffset + dataoffset;
                    f.dataSize = datasize;
                    f.name = name;
                    f.fullName = fullname;
                    f.data = null;

                    fileEntries.put(pathToKey(fullname), f);
                    parentdir.childrenFiles.add(f);
                }
            }
        }
    }
    
    // fix: ignore the root directory name in filenames
    // SMG ignores it as well, and some RARC packers set it arbitrarily
    private String pathToKey(String path)
    {
        String ret = path.toLowerCase();
        ret = ret.substring(1);
        if (ret.indexOf("/") == -1) return "/";
        ret = ret.substring(ret.indexOf("/"));
        return ret;
    }
    
    private int align32(int val)
    {
        return (val + 0x1F) & ~0x1F;
    }
    
    private int dirMagic(String name)
    {
        String uppername = name.toUpperCase();
        int ret = 0;
        
        for (int i = 0; i < 4; i++)
        {
            ret <<= 8;
            
            if (i >= uppername.length())
                ret += 0x20;
            else
                ret += uppername.charAt(i);
        }
        
        return ret;
    }
    
    private short nameHash(String name)
    {
        short ret = 0;
        for (char ch : name.toCharArray())
        {
            ret *= 3;
            ret += ch;
        }
        return ret;
    }
    
    @Override
    public void save() throws IOException
    {
        for (FileEntry fe : fileEntries.values())
        {
            if (fe.data != null) continue;
            file.position(fe.dataOffset);
            fe.data = file.readBytes(fe.dataSize);
        }
        
        int dirOffset = 0x40;
        int fileOffset = dirOffset + align32(dirEntries.size() * 0x10);
        int stringOffset = fileOffset + align32((fileEntries.size() + (dirEntries.size() * 3) - 1) * 0x14);
        
        int dataOffset = stringOffset;
        int dataLength = 0;
        for (DirEntry de : dirEntries.values())
            dataOffset += de.name.length() + 1;
        for (FileEntry fe : fileEntries.values())
        {
            dataOffset += fe.name.length() + 1;
            dataLength += align32(fe.dataSize);
        }
        dataOffset += 5;
        dataOffset = align32(dataOffset);
        
        int dirSubOffset = 0;
        int fileSubOffset = 0;
        int stringSubOffset = 0;
        int dataSubOffset = 0;
        
        file.setLength(dataOffset + dataLength);
        
        // RARC header
        // certain parts of this will have to be written later on
        file.position(0);
        file.writeInt(0x52415243);
        file.writeInt(dataOffset + dataLength);
        file.writeInt(0x00000020);
        file.writeInt(dataOffset - 0x20);
        file.writeInt(dataLength);
        file.writeInt(dataLength);
        file.writeInt(0x00000000);
        file.writeInt(0x00000000);
        file.writeInt(dirEntries.size());
        file.writeInt(dirOffset - 0x20);
        file.writeInt(fileEntries.size() + (dirEntries.size() * 3) - 1);
        file.writeInt(fileOffset - 0x20);
        file.writeInt(dataOffset - stringOffset);
        file.writeInt(stringOffset - 0x20);
        file.writeInt(unk38);
        file.writeInt(0x00000000);
        
        file.position(stringOffset);
        file.writeString("ASCII", ".", 0);
        file.writeString("ASCII", "..", 0);
        stringSubOffset += 5;
        
        Stack<Iterator<DirEntry>> dirstack = new Stack<>();
        Object[] entriesarray = dirEntries.values().toArray();
        DirEntry curdir = (DirEntry)entriesarray[0];
        int c = 1;
        while (curdir.parentDir != null) curdir = (DirEntry)entriesarray[c++];
        short fileid = 0;
        for (;;)
        {
            // write the directory node
            curdir.tempID = dirSubOffset / 0x10;
            file.position(dirOffset + dirSubOffset);
            file.writeInt((curdir.tempID == 0) ? 0x524F4F54 : dirMagic(curdir.name));
            file.writeInt(stringSubOffset);
            file.writeShort(nameHash(curdir.name));
            file.writeShort((short)(2 + curdir.childrenDirs.size() + curdir.childrenFiles.size()));
            file.writeInt(fileSubOffset / 0x14);
            dirSubOffset += 0x10;
            
            if (curdir.tempID > 0)
            {
                file.position(curdir.tempNameOffset);
                file.writeShort((short)stringSubOffset);
                file.writeInt(curdir.tempID);
            }
            file.position(stringOffset + stringSubOffset);
            stringSubOffset += file.writeString("ASCII", curdir.name, 0);
            
            // write the child file/dir entries
            file.position(fileOffset + fileSubOffset);
            for (DirEntry cde : curdir.childrenDirs)
            {
                file.writeShort((short)0xFFFF);
                file.writeShort(nameHash(cde.name));
                file.writeShort((short)0x0200);
                cde.tempNameOffset = (int)file.position();
                file.skip(6);
                file.writeInt(0x00000010);
                file.writeInt(0x00000000);
                fileSubOffset += 0x14;
            }
            
            for (FileEntry cfe : curdir.childrenFiles)
            {
                file.position(fileOffset + fileSubOffset);
                file.writeShort(fileid);
                file.writeShort(nameHash(cfe.name));
                file.writeShort((short)0x1100);
                file.writeShort((short)stringSubOffset);
                file.writeInt(dataSubOffset);
                file.writeInt(cfe.dataSize);
                file.writeInt(0x00000000);
                fileSubOffset += 0x14;
                fileid++;

                file.position(stringOffset + stringSubOffset);
                stringSubOffset += file.writeString("ASCII", cfe.name, 0);
                
                file.position(dataOffset + dataSubOffset);
                cfe.dataOffset = (int)file.position();
                byte[] thedata = Arrays.copyOf(cfe.data, cfe.dataSize);
                file.writeBytes(thedata);
                dataSubOffset += align32(cfe.dataSize);
                cfe.data = null;
            }
            
            file.position(fileOffset + fileSubOffset);
            file.writeShort((short)0xFFFF);
            file.writeShort((short)0x002E);
            file.writeShort((short)0x0200);
            file.writeShort((short)0x0000);
            file.writeInt(curdir.tempID);
            file.writeInt(0x00000010);
            file.writeInt(0x00000000);
            file.writeShort((short)0xFFFF);
            file.writeShort((short)0x00B8);
            file.writeShort((short)0x0200);
            file.writeShort((short)0x0002);
            file.writeInt((curdir.parentDir != null) ? curdir.parentDir.tempID : 0xFFFFFFFF);
            file.writeInt(0x00000010);
            file.writeInt(0x00000000);
            fileSubOffset += 0x28;
            
            // determine who's next on the list
            // * if we have a child directory, process it
            // * otherwise, look if we have remaining siblings
            // * and if none, go back to our parent and look for siblings again
            // * until we have done them all
            if (!curdir.childrenDirs.isEmpty())
            {
                dirstack.push(curdir.childrenDirs.iterator());
                curdir = dirstack.peek().next();
            }
            else
            {
                curdir = null;
                while (curdir == null)
                {
                    if (dirstack.empty())
                        break;
                    
                    Iterator<DirEntry> it = dirstack.peek();
                    if (it.hasNext())
                        curdir = it.next();
                    else
                        dirstack.pop();
                }
                
                if (curdir == null)
                    break;
            }
        }
        
        file.save();
    }

    @Override
    public void close() throws IOException
    {
        file.close();
    }


    @Override
    public boolean directoryExists(String directory)
    {
        return dirEntries.containsKey(pathToKey(directory));
    }

    @Override
    public List<String> getDirectories(String directory)
    {
        if (!dirEntries.containsKey(pathToKey(directory))) 
            return null;
        
        DirEntry dir = dirEntries.get(pathToKey(directory));
        
        List<String> ret = new ArrayList<>();
        for (DirEntry de : dir.childrenDirs)
        {
            ret.add(de.name);
        }
        
        return ret;
    }


    @Override
    public boolean fileExists(String filename)
    {
        return fileEntries.containsKey(pathToKey(filename));
    }

    @Override
    public List<String> getFiles(String directory)
    {
        if (!dirEntries.containsKey(pathToKey(directory))) 
            return null;
        
        DirEntry dir = dirEntries.get(pathToKey(directory));
        
        List<String> ret = new ArrayList<>();
        for (FileEntry fe : dir.childrenFiles)
        {
            ret.add(fe.name);
        }
        
        return ret;
    }

    @Override
    public FileBase openFile(String filename) throws FileNotFoundException
    {
        if (!fileEntries.containsKey(pathToKey(filename)))
            throw new FileNotFoundException(filename + " not found in RARC");
        
        try
        {
            return new RarcFile(this, filename);
        }
        catch (IOException ex)
        {
            throw new FileNotFoundException("got IOException");
        }
    }
    
    @Override
    public void createFile(String parent, String newfile)
    {
        String parentkey = pathToKey(parent);
        String fnkey = pathToKey(parent + "/" + newfile);
        if (!dirEntries.containsKey(parentkey)) return;
        if (fileEntries.containsKey(fnkey)) return;
        if (dirEntries.containsKey(fnkey)) return;
        DirEntry parentdir = dirEntries.get(parentkey);
        
        FileEntry fe = new FileEntry();
        fe.data = new byte[0];
        
        fe.dataSize = fe.data.length;
        fe.fullName = parent + "/" + newfile;
        fe.name = newfile;
        fe.parentDir = parentdir;
        
        parentdir.childrenFiles.add(fe);
        fileEntries.put(pathToKey(fe.fullName), fe);
    }

    @Override
    public void renameFile(String file, String newname)
    {
        file = pathToKey(file);
        if (!fileEntries.containsKey(file)) return;
        FileEntry fe = fileEntries.get(file);
        DirEntry parent = fe.parentDir;
        
        String parentkey = pathToKey(parent.fullName + "/" + newname);
        if (fileEntries.containsKey(parentkey) ||
            dirEntries.containsKey(parentkey)) 
            return;
        
        String fnkey = pathToKey(fe.fullName);
        fileEntries.remove(fnkey);
        
        fe.name = newname;
        fe.fullName = parent.fullName + "/" + newname;
        
        fileEntries.put(fnkey, fe);
    }

    @Override
    public void deleteFile(String file)
    {
        file = pathToKey(file);
        if (!fileEntries.containsKey(file)) return;
        FileEntry fe = fileEntries.get(file);
        DirEntry parent = fe.parentDir;
        
        parent.childrenFiles.remove(fe);
        fileEntries.remove(file);
    }


    // support functions for RarcFile
    public byte[] getFileContents(String fullname) throws IOException
    {
        FileEntry fe = fileEntries.get(pathToKey(fullname));

        file.position(fe.dataOffset);
        return file.readBytes((int)fe.dataSize);
    }

    public void reinsertFile(RarcFile _file) throws IOException
    {
        FileEntry fe = fileEntries.get(pathToKey(_file.fileName));
        fe.data = _file.getContents();
        fe.dataSize = (int)_file.getLength();
    }


    private class FileEntry
    {
        public FileEntry()
        { 
            data = null;
        }
        
        public int dataOffset;
        public int dataSize;

        public DirEntry parentDir;

        public String name;
        public String fullName;
        
        public byte[] data;
    }

    private class DirEntry
    {
        public DirEntry()
        {
            childrenDirs = new LinkedList<>();
            childrenFiles = new LinkedList<>();
        }

        public DirEntry parentDir;
        public LinkedList<DirEntry> childrenDirs;
        public LinkedList<FileEntry> childrenFiles;

        public String name;
        public String fullName;
        
        public int tempID;
        public int tempNameOffset;
    }


    private FileBase file;

    private int unk38;

    private LinkedHashMap<String, FileEntry> fileEntries;
    private LinkedHashMap<String, DirEntry> dirEntries;
}
